### ISSUE TITLE
Double clicking to toggle fullscreen always (briefly) pauses/unpauses the playback

### ISSUE THREAD

**Issue by OscarL** [clsid2/mpc-hc#1218]
MPC-HC version: 1.9.14
OS: Win10 1909 64 bits

No matter how fast I double click to toggle fullscreen, it seems that the "left click to (un)pause" gets executed before the fullscreen toggle, so an unnecessary pause/un-pause is introduced.

I would expect that entering and exiting fullscreen via double click with the left mouse button should not disrupt the playback.

It is of course a small annoyance, I don't even mind the pause of VIDEO (after all, toggling fullscreen does disrupt the visuals momentarily), but the pause on the audio tend to get on my nerves :-)

Any way, if that's my "mayor problem" with MPC-HC.... it's because it is otherwise a joy to use!

Thanks a lot for keeping alive such a nice program!

**Comment by clsid2** [comment]
When you double click there are four actions triggered by Windows:
Left down, Left up, Left Double, Left Up.
Pause is assigned to Left Up. You can remove that assignment, but you would loose the ability to pause by left clicking on the video.

You can also toggle fullscreen by pressing F11.

**Comment by OscarL** [comment]
So... a "WontFix", I guess. Bummer.

> When you double click there are four actions triggered by Windows:
> Left down, Left up, Left Double, Left Up.

Still, other software are capable of detecting a double click, and avoid the spurious single clicks.
Measuring the time between the first mouse down and the next, if < to, say 200 ms, then it's a double click.

I remember that in Delphi at least there were even separate events for mouse down/up, plus one specific for double clicks. That's why I thought it was a bug and not the intended behavior (nonsensical as it is).

> Pause is assigned to Left Up. You can remove that assignment, but you would loose the ability to pause by left clicking on the video.
> You can also toggle fullscreen by pressing F11.

Thanks for the suggestions. None that I'm particularly fond of... but... it's your app, and you make the calls.

I hope you reconsider it (maybe if more people ask for it?).

Thanks again for your work. It is appreciated.

**Comment by clsid2** [comment]
The single click happens before it becomes a double-click. That is how it works in MFC window classes. So only way to avoid it would be to delay first click (for up to 500ms!) and use a timer to process it if no second click occurs. That solution is very ugly so I am not going to do that.
Another possibility would to to ignore single-clicks in a specific area of the screen. But I am not really motivated to implement such hidden workarounds.
There might be the ability to customize the toolbar buttons in the future. Then having a fullscreen button could be a possibility.

You can also just get a mouse that has extra buttons on the side/middle. Assign fullscreen to one of those buttons if you don't have a keyboard nearby.

**Comment by OscarL** [comment]
I understand you not wanting to introduce a nasty workaround (specially that ugly "ignore single clicks in *this* area).

But...

> So only way to avoid it would be to delay first click (for up to 500ms!)

Is that from the first mouse-down to the second mouse-up? Maybe a smaller threshold between first mouse-up and second mouse down then?

I wonder how Firefox+Youtube do it. Pause/Unpause on mouse up, fullscreen toggle with double click. That's the more popular combo I can think of that behaves like that (it would be nice to have a consistent behavior across the board, of course).

Anyway... As I'm not able to contribute code to fix to this, I'll just keep trying to get used to it :-)

Have a great day!

**Comment by clsid2** [comment]
https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-nclbuttondblclk

And I clearly see the exact same behavior here with Firefox+Youtube. It also does a play/pause.

**Comment by OscarL** [comment]
This will be my last comment regarding this issue, as I don't want to bother you any longer with this.

> And I clearly see the exact same behavior here with Firefox+Youtube. It also does a play/pause.

I just triple checked (Firefox 90, Win 1909). It does NOT pauses/un-pauses the audio while toggling fullscreen via double click **IF** the double click is fast enough.

I have to do a deliberately slow double click for the pause/unpause to interrupt the audio playback, so... it seems that my usual double click delay is beneath Firefox's threshold for double click detection, and thus, I've assumed that was the case for everyone.

Oh, well. Too bad for me, I guess :-)

**Comment by clsid2** [comment]
Well, I clearly briefly see a pause icon even when double-clicking fast on youtube.
And when single clicking to pause you will notice a delay before the audio pauses. So it probably uses a buffer for the audio.

Fact remains that Firefox is a totally different program and you can NOT compare its behavior with MPC-HC.

**Comment by bobdig** [comment]
@clsid2 I also find this annoying. Reading this here I kinda getting it why it is working for VLC and not for MPC-HC. But if you would find a way around that, it would be a much better experience.

**Comment by clsid2** [comment]
It is IMPOSSIBLE to change/fix this behavior when something is assigned to both single click and double-click.

Use F11 for fullscreen, or remove the single mouse click assignment for pause.

**Comment by OscarL** [comment]
@bobdig I was just testing QMPlay2 on a Haiku installation of mine, and I came across a setting that basically meant: "use left click to pause/un-pause, but only after a small delay, so you can double click to toggle full-screen mode", and I intermediately remembered opening this issue.

I DO totally understand @clsid2 position. There is no "clean" way to fix this, other than introducing a nasty delay/wait thread/threshold. I think I understand his position from a programmer view.

Still, as an ex programmer, an ex QA Team Lead. and as an everyday user... users, in the end, care much more for the [Principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) than they do about the inner working of software.

In the end, it's up to clsid2's decision, as it should be (considering that I'm "only complaining, and not able to contribute code", I understand, and respect that).

I just hope he understand that this type of issues are NOT a critique on him, or the excellent and welcome work he does on keeping this software alive. They are just a reflection of what an everyday user considers to be the expected behavior... (that as the Wikipedia article implies... not always makes sense, but humans rarely do :-) ).

Edit: for the record... I now use "f" to toggle fullscreen on MPC-HC.... still trying to gain some muscle memory. Still wish I was just able to double click without an "audio glitch" :-).

**Comment by fuzzzeee** [comment]
I would also love double-click support in keys configuration!

I absolutely love MPC but the pause when toggling fullscreen is the only slight niggle I have with it

It's not a major issue of course - just my 2 cents 😄

**Comment by kkdexz** [comment]
> I would also love double-click support in keys configuration!

What do you mean? There already is double-click support in hotkey settings. Just scroll to the right in the "Keys" options. And not only left click, but pretty much all mouse buttons (up to 5). For e.g., I have right double-click set to open options.

**Comment by bibsp** [comment]
It was always a needed feature in MPC. Happy to get it finally.
Thanks **clsid2** for the implementation. Working perfectly as it should. ❤

**Comment by OscarL** [comment]
Very nice!

Installed latest dev version (MPC-HC.1.9.24.8.x64.exe), set Window's double-click time to its minimum wait time (to minimize the delay)... and now everything works the way I like it the most :-D

Thanks a lot @clsid2!
